x/mongo/driver/operation.go

```diff
@@ -1078,6 +1094,28 @@ func (op Operation) addServerAPI(dst []byte) []byte {
 	return dst
 }
 
+// addMaxTimeMS adds a 'maxTimeMS' field to the wire message in dst based on the current context's deadline
+// and the 90th percentile RTT if op.Timeout is set.
+func (op Operation) addMaxTimeMS(ctx context.Context, dst []byte, desc description.SelectedServer) []byte {
+	if op.Timeout != nil {
+		// Retrieve current deadline from context, convert it to timeout and append
+		// maxTimeMS as (timeout - 90th percentile RTT).
+		if deadline, deadlineSet := ctx.Deadline(); deadlineSet {
+			remainingTimeout := time.Until(deadline)
+			// If 90th percentile RTT is less than remainingTimeout, error instead
+			// of attempting to send message to server.
+			// TODO(CSOT): Uncomment this once RTT90 is actually being calculated.
+			// if desc.RTT90 < remainingTimeout {
+			//	return dst, internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
+			//		"Remaining timeout %v applied from Timeout is less than 90th percentile RTT", remainingTimeout)
+			// }
```

Comment:
Consider applying this logic where the existing short-circuit logic is currently: [operation.go#L496-497](https://github.com/mongodb/mongo-go-driver/blob/9101b419216336cec091fc70d1f622efa12f5ca0/x/mongo/driver/operation.go#L496-L497). Also, this should replace that existing short-circuit logic.